Comparing certain objects throws `TypeError: Cannot convert object to primitive value`

When comparing objects with certain `length` properties, the reporter throws when stringifying the diff.  Reduced repro:
```javascript
it('compare two objects', function () {
   assert.deepEqual({length: {toString: 0}}, {length: 1});
});
```
Expected:
```diff
AssertionError: { length: { toString: 0 } } deepEqual { length: 1 }
+ expected
- actual

 {
-  "length": {
-    "toString": 0
-  }
+  "length": 1
 }
```
Actual:
```
TypeError: Cannot convert object to primitive value
at jsonStringify (mocha/lib/utils.js:504:7)
at Object.exports.stringify (mocha/lib/utils.js:431:14)
at mocha/lib/reporters/base.js:203:37
at Array.forEach (native)
at Function.exports.list (mocha/lib/reporters/base.js:163:12)
at Spec.Base.epilogue (mocha/lib/reporters/base.js:328:10)
at Runner.emit (events.js:129:20)
at mocha/lib/runner.js:791:12
at mocha/lib/runner.js:644:9
at next (mocha/lib/runner.js:284:14)
```

--- a/lib/utils.js
+++ b/lib/utils.js
@@ -454,7 +454,7 @@ function jsonStringify(object, spaces, depth) {
   var space = spaces * depth;
   var str = isArray(object) ? '[' : '{';
   var end = isArray(object) ? ']' : '}';
-  var length = object.length || exports.keys(object).length;
+  var length = typeof object.length === 'number' ? object.length : exports.keys(object).length;
   // `.repeat()` polyfill
   function repeat(s, n) {
     return new Array(n).join(s);

--- a/test/acceptance/utils.js
+++ b/test/acceptance/utils.js
@@ -340,6 +340,10 @@ describe('lib/utils', function () {
         stringify({symbol: symbol}).should.equal('{\n  "symbol": Symbol(value)\n}')
       });
     }
+
+    it('should handle length properties that cannot be coerced to a number', function () {
+      stringify({length: {toString: 0}}).should.equal('{\n  "length": {\n    "toString": 0\n  }\n}');
+    });
   });
 
   describe('type', function () {

